torch/_logging/internal.py

```diff
@@ -0,0 +1,384 @@
+import collections
+import itertools
+import logging
+import functools
+import os
+import re
+from importlib import __import__
+from typing import DefaultDict, Dict, Set
+from dataclasses import dataclass, field
+
+DEFAULT_LOG_LEVEL = logging.WARN
+DEFAULT_FORMATTER = logging.Formatter(
+    "[%(asctime)s] %(name)s: [%(levelname)s] %(message)s"
+)
+
+@dataclass
+class LogRegistry:
+    # shorthand name to log qualified name
+    name_to_log_qname: Dict[str, str] = field(default_factory=dict)
+    # shorthand name to record type
+    name_to_rec_type: Dict[str, type] = field(default_factory=dict)
+    # log qualified name to set of artifact types
+    log_qname_to_rec_types: Dict[str, Set[type]] = field(default_factory=dict)
+    # log qualified name to all supported shorthand names
+    log_qname_to_loggable_names: Dict[str, Set[str]] = field(default_factory=dict)
+    # shorthand names of logs which support verbosity
+    log_names_with_verbosity: Set[str] = field(default_factory=set)
+
+    def is_artifact(self, name):
+        return name in self.name_to_rec_type
+
+    def is_log(self, name):
+        return name in self.name_to_log_qname and name not in self.name_to_rec_type
+
+    def register_log(self, setting_name, log_qname, has_verbosity):
+        # Check dupes
+        assert log_qname not in self.log_qname_to_rec_types
+        assert log_qname not in self.log_qname_to_loggable_names
+        assert setting_name not in self.name_to_log_qname
+
+        self.name_to_log_qname[setting_name] = log_qname
+        self.log_qname_to_rec_types[log_qname] = set()
+        self.log_qname_to_loggable_names[log_qname] = set()
+
+        if has_verbosity:
+            self.log_names_with_verbosity.add(setting_name)
+
+
+    def register_artifact(self, setting_name, artifact_type, log_qname, off_by_default):
+        self.log_qname_to_loggable_names[log_qname].add(setting_name)
+        self.name_to_log_qname[setting_name] = log_qname
+        self.name_to_rec_type[setting_name] = artifact_type
+        # if off by default, don't enable it
+        # when log_name's log_level is set to DEBUG
+        if not off_by_default:
+            self.log_qname_to_rec_types[log_qname].add(artifact_type)
+
+    def get_loggable_names(self):
+        return list(itertools.chain(self.name_to_log_qname.keys(), self.name_to_rec_type.keys()))
+
+    def get_log_qnames(self):
+        return set(self.log_qname_to_rec_types.keys())
+
+    def supports_verbosity(self, log_name):
+        return log_name in self.log_names_with_verbosity
+
+    def register_existing_child_log(self, parent_qname, log_qname):
+        self.name_to_log_qname[log_qname] = log_qname
+        self.log_qname_to_rec_types = set(self.log_qname_to_rec_types[parent_qname])
+
+log_registry = LogRegistry()
```

Comment:
nit: globals get all caps

Reply:
I thought this was only if they are constants?